File: compiler/bench/graycode.c
// expected: 5000000
// Gray code - convert binary to Gray code and back
#include <stdio.h>
#include <stdint.h>

uint32_t to_gray(uint32_t n) {
    return n ^ (n >> 1);
}

uint32_t from_gray(uint32_t gray) {
    uint32_t n = gray;
    while (gray >>= 1) {
        n ^= gray;
    }
    return n;
}

int main(void) {
    long count = 0;
    for (int i = 0; i < 5000000; i++) {
        if (from_gray(to_gray(i)) == i) count++;
    }
    printf("%ld\n", count);
    return 0;
}

File: compiler/bench/graycode.fs
\ expected: 5000000
\ Gray code - convert binary to Gray code and back

: to-gray ( n -- gray )
  dup 1 rshift xor ;

: from-gray ( gray -- n )
  dup
  begin dup 1 rshift dup while
    rot xor swap
  repeat drop ;

: main
  0
  5000000 0 do
    i to-gray from-gray
    i = if 1+ then
  loop
  . cr ;
